pow/src/compute/mod.rs

```diff
@@ -89,34 +92,82 @@ fn need_new_vm<M: randomx::WithCacheMode>(
 	need_new_vm
 }
 
+fn do_new_vm<M: randomx::WithCacheMode>(
+	key_hash: &H256,
+	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
+	shared_caches: &Mutex<LruCache<H256, Arc<randomx::Cache<M>>>>,
+	f_has_large_pages: fn(&Config) -> bool,
+) -> Result<(), ()> {
+	let mut shared_caches = shared_caches.lock();
+
+	if !f_has_large_pages(global_config()) {
+		if let Some(cache) = shared_caches.get_mut(key_hash) {
+			machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+
+			Ok(())
+		} else {
+			info!(
+				target: "kulupu-randomx",
+				"At block boundary, generating new RandomX {} cache with key hash {} ...",
+				M::description(),
+				key_hash,
+			);
+
+			let cache = match randomx::Cache::new(&key_hash[..], global_config()) {
+				Ok(cache) => Arc::new(cache),
+				Err(_) => panic!("Cache allocation failed"),
+				};
+
+			shared_caches.insert(*key_hash, cache.clone());
+			machine.replace(Some((*key_hash, randomx::VM::new(cache, global_config()))));
+
+			Ok(())
+		}
+	} else {	
+		if let Some(cache) = shared_caches.get_mut(key_hash) {
+			machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+
+			Ok(())
+		} else {
+			machine.replace(None);
+			let key_to_replace = match (*shared_caches)
+					.iter()
+					.find(|&(_, cache)| Arc::strong_count(cache) == 1)
+					.and_then(|(key, _)| Some(*key)) {
+				Some(key) => key,
+				None => return Err(()),
```

Comment:
This means that if large pages is enabled, during epoch switches the miner may experience large hashrate drop, but I guess this is a reasonable tradeoff.